dht_bootstrap.go

```diff
@@ -132,51 +117,24 @@ func (dht *IpfsDHT) runBootstrap(ctx context.Context, cfg BootstrapConfig) error
 	defer bslog("end")
 	defer log.EventBegin(ctx, "dhtRunBootstrap").Done()
 
-	var merr u.MultiErr
-
-	randomID := func() peer.ID {
-		// 16 random bytes is not a valid peer id. it may be fine becuase
-		// the dht will rehash to its own keyspace anyway.
-		id := make([]byte, 16)
-		rand.Read(id)
-		id = u.Hash(id)
-		return peer.ID(id)
-	}
-
-	// bootstrap sequentially, as results will compound
-	runQuery := func(ctx context.Context, id peer.ID) {
+	doQuery := func(n int, target string, f func(context.Context) error) error {
+		log.Debugf("Bootstrapping query (%d/%d) to %s", n, cfg.Queries, target)
 		ctx, cancel := context.WithTimeout(ctx, cfg.Timeout)
 		defer cancel()
-
-		p, err := dht.FindPeer(ctx, id)
-		if err == routing.ErrNotFound {
-			// this isn't an error. this is precisely what we expect.
-		} else if err != nil {
-			merr = append(merr, err)
-		} else {
-			// woah, actually found a peer with that ID? this shouldn't happen normally
-			// (as the ID we use is not a real ID). this is an odd error worth logging.
-			err := fmt.Errorf("Bootstrap peer error: Actually FOUND peer. (%s, %s)", id, p)
-			log.Warningf("%s", err)
-			merr = append(merr, err)
-		}
+		return f(ctx)
 	}
 
-	// these should be parallel normally. but can make them sequential for debugging.
-	// note that the core/bootstrap context deadline should be extended too for that.
-	for i := 0; i < cfg.Queries; i++ {
-		id := randomID()
-		log.Debugf("Bootstrapping query (%d/%d) to random ID: %s", i+1, cfg.Queries, id)
-		runQuery(ctx, id)
+	// Do all but one of the bootstrap queries as random walks.
+	for i := 1; i < cfg.Queries; i++ {
```

Comment:
This is not going to play well with the default value of `Queries` which is 1 (see `DefaultBootstrapConfig`). We'll always end up querying only ourselves (i.e. bypassing this for). I suggest we return the loop range to the original (starting with `i = 0`).